src/crimson/os/seastore/extentmap_manager/btree/btree_extentmap_manager.h

```diff
@@ -0,0 +1,63 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <iostream>
+
+#include <boost/intrusive_ptr.hpp>
+#include <boost/smart_ptr/intrusive_ref_counter.hpp>
+#include <seastar/core/future.hh>
+
+#include "include/ceph_assert.h"
+#include "crimson/osd/exceptions.h"
+
+#include "crimson/os/seastore/extentmap_manager.h"
+#include "crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node.h"
+#include "crimson/os/seastore/seastore_types.h"
+#include "crimson/os/seastore/transaction_manager.h"
+
+namespace crimson::os::seastore::extentmap_manager {
+/**
+ * BtreeExtentMapManager
+ *
+ * Uses a btree to track :
+ * objaddr_t -> laddr_t mapping for each onode extentmap
+ *
+ */
+
+class BtreeExtentMapManager : public ExtentMapManager {
+  TransactionManager &tm;
+
+  /* get_extmap_root
+   *
+   * load extent map tree root node
+   */
+  using get_root_ertr = TransactionManager::read_extent_ertr;
+  using get_root_ret = get_root_ertr::future<ExtMapNodeRef>;
+  get_root_ret get_extmap_root(extmap_root_t &extmap_root, Transaction &t);
+
+  using insert_lextent_ertr = TransactionManager::read_extent_ertr;
+  using insert_lextent_ret = insert_lextent_ertr::future<extent_mapping_t >;
+  insert_lextent_ret insert_lextent(extmap_root_t &extmap_root, Transaction &t,
+                                    ExtMapNodeRef extent, objaddr_t lo,
+				    lext_map_val_t val);
+
+public:
+  explicit BtreeExtentMapManager(TransactionManager &tm);
+
+  initialize_extmap_ret initialize_extmap(Transaction &t) final;
+
+  find_lextent_ret find_lextent(extmap_root_t &extmap_root, Transaction &t, objaddr_t lo, extent_len_t len) final;
+
+  add_lextent_ret add_lextent(extmap_root_t &extmap_root, Transaction &t, objaddr_t lo, lext_map_val_t val) final;
```

Comment:
For these two, it'll be rare that extmap_root is modified.  How is the caller supposed to work out whether you modified it?  They'll need to know because they'll have to somehow persist the change (probably just mark the onode dirty or w/e).

Reply:
add a state in extmap_root_t, when extmap_root is changed, the state is set to MUTATED.  then caller can check the state then get to know extmap root address is changed.